compiler/base.ts

```diff
@@ -71,6 +74,100 @@ export type CompilerOptions = {
 
 export type CompileResult<T> = [T, ParseWarning | null];
 
+class GlobalBlock extends Block {
+  kind = 'global';
+}
+
+class ProgramBlock extends Block {
+  kind = 'program';
+}
+
+class CommandBlock extends Block {
+  kind = 'command';
+}
+
+export function isRootBlock(block: Block): boolean {
+  return (
+    block instanceof ProgramBlock ||
+    block instanceof CommandBlock ||
+    block instanceof GlobalBlock
+  );
+}
+
+//
+// if, else, else if and endif
+//
+
+class IfBlock extends Block {
+  kind = 'if';
+
+  constructor(public elseJump: Short) {
+    super();
+  }
+
+  visitElseInstr(else_: Else): void {
+    const endJump = this.compiler.else_(this.elseJump);
+    this.next(else_, new ElseBlock(endJump));
+  }
+
+  visitElseIfInstr(elseIf: ElseIf): void {
+    const endJump = this.compiler.else_(this.elseJump);
+    const elseJump = this.compiler.if_(elseIf.condition);
+    this.next(elseIf, new ElseIfBlock(elseJump, endJump));
+  }
+
+  visitEndIfInstr(_endIf: EndIf): void {
+    // TODO: Optimize for no 'else'
+    const endJump = this.compiler.else_(this.elseJump);
+    this.compiler.endIf(endJump);
+    this.end();
+  }
+}
+
+class ElseBlock extends Block {
+  kind = 'else';
+
+  constructor(public endJump: Short) {
+    super();
+  }
+
+  visitEndIfInstr(_endIf: EndIf): void {
+    this.compiler.endIf(this.endJump);
+    this.end();
+
+    let block: Block | null = this.previous;
+    while (block instanceof ElseIfBlock) {
```

Comment:
Handling `IfBlock` is unnecessary because the end jump is attached to the following `ElseIf`, right? Let's check that.

Reply:
I think the fact that the test shows a sensible amount of jumping, that we can be confident this is being handled correctly.